File: lc1008.cs
public class Solution {
    public TreeNode BstFromPreorder(int[] preorder) {
        return ConstructBST(preorder, 0, preorder.Length-1);
    }
    private TreeNode ConstructBST(int[] preorder, int start, int end)
    {
        if(start<0 || start>end)
            return null;   
        TreeNode t = new TreeNode(preorder[start]);        
        int endIdx = FindIdxofElementGTX(preorder[start],preorder, start+1, end);
        if(endIdx ==-1)
        {
            t.left = ConstructBST(preorder, start+1, end);
            //t.right = ConstructBST(preorder, endIdx, end);
        }
        else
        {
            t.left = ConstructBST(preorder, start+1, endIdx-1);
            t.right = ConstructBST(preorder, endIdx, end);
        }
        
        
        return t;
    }
    private int FindIdxofElementGTX(int x, int[] preorder, int start, int end)
    {
        for(int i=start;i<=end;i++)
        {
            if(preorder[i]>x)
                return i;
        }
        return -1;
    }
}

File: lc1008.java
class Solution {
    public TreeNode bstFromPreorder(int[] preorder) {
        int n= preorder.length;
        if(n==0)
            return null;
        return ConstructBst(preorder,0,n-1);
    }
    private TreeNode ConstructBst(int[] preorder, int start, int end)
    {
        if(start>end)
            return null;
        TreeNode root = new TreeNode(preorder[start]);
        int idx = findRightRootIndex(preorder, preorder[start], start+1, end);
        if(idx!=-1)
        {
            root.left = ConstructBst(preorder, start+1, idx-1);
            root.right = ConstructBst(preorder, idx, end);
        }
        else
        {
            root.left = ConstructBst(preorder, start+1, end);
        }
        return root;
    }
    private int findRightRootIndex(int[] preorder, int target, int start, int end)
    {
        int idx = -1;
        for(int i=start;i<=end;i++)
        {
            if(preorder[i]<target)
                continue;
            else
            {
                idx = i;
                break;
            }
        }
        return idx;
    }
}
